privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+        return "push"
+
+    @staticmethod
+    def get_class_prefix():
+        return "PIPU"
+
+    @staticmethod
+    def get_class_info(key=None, ret='all'):
+        """
+        returns all or a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: str
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+
+        :return: subsection if key exists or user defined
+        :rtype : s.o.
+        """
+        gws = get_smsgateway(gwtype=u'privacyidea.lib.smsprovider.FirebaseProvider.FirebaseProvider')
+        res = {'type': 'push',
+               'title': _('PUSH Token'),
+               'description':
+                    _('PUSH: Send a push notification to a smartphone.'),
+               'user': ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       PUSH_ACTION.FIREBASE_CONFIG: {
+                           'type': 'str',
+                           'desc': _('The configuration of your Firebase application.'),
+                           'group': "PUSH",
+                           'value': [gw.identifier for gw in gws]
+                       },
+                   },
+                   SCOPE.AUTH: {
+                       PUSH_ACTION.MOBILE_TEXT: {
+                           'type': 'str',
+                           'desc': _('The question the user sees on his mobile phone.'),
+                           'group': 'PUSH'
+                       },
+                       PUSH_ACTION.MOBILE_TITLE: {
+                           'type': 'str',
+                           'desc': _('The title of the notification, the user sees on his mobile phone.'),
+                           'group': 'PUSH'
+                       }
+                   }
+               },
+        }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+
+        return ret
+
+    @log_with(log)
+    def update(self, param, reset_failcount=True):
+        """
+        process the initialization parameters
+
+        We need to distinguish the first authentication step
+        and the second authentication step.
+
+        1. step:
+            parameter type contained.
+            parameter genkey contained.
+
+        2. step:
+            parameter serial contained
+            parameter fbtoken contained
+            parameter pubkey contained
+
+        :param param: dict of initialization parameters
+        :type param: dict
+
+        :return: nothing
+        """
+        upd_param = {}
+        for k, v in param.items():
+            upd_param[k] = v
+
+        if "serial" in upd_param and "fbtoken" in upd_param and "pubkey" in upd_param:
+            # We are in step 2:
+            if not self.token.rollout_state == "clientwait":
```

Comment:
Hm, I find ``!=`` more readable than ``not ... == ``. What do you think?

Reply:
will do

Reply:
done in ee1593b9